source/bdDetect.py

```diff
@@ -238,121 +237,119 @@ class Detector(object):
-					# unless a display has been found.
-					return
-				braille._BgThread.queueApc(self._BgScanApc)
+		self._detectUsb = usb
+		self._detectBluetooth = bluetooth
+		self._limitToDevices = limitToDevices
+		if self._queuedFuture:
+			# This will cancel a queued scan (i.e. not the currently running scan, if any)
+			# If this future belongs to a scan that is currently running or finished, this does nothing.
+			self._queuedFuture.cancel()
+		self._queuedFuture = self._executor.submit(self._bgScan, usb, bluetooth, limitToDevices)
 
 	def _stopBgScan(self):
 		"""Stops the current scan as soon as possible and prevents a queued scan to start."""
-		if not self._runningApcLock.locked():
-			# No scan to stop
-			return
 		self._stopEvent.set()
-		self._scanQueuedSafe = False
-
-	def _bgScan(self, param):
-		if self._runningApcLock.locked():
-			log.debugWarning("Braille display detection background scan APC executed while one is already running")
-			return
-		with self._runningApcLock:
-			while self._scanQueuedSafe:
-				# Clear the stop event before a scan is started.
-				# Since a scan can take some time to complete, another thread can set the stop event to cancel it.
-				self._stopEvent.clear()
-				with self._queuedScanLock:
-					self._scanQueued = False
-					detectUsb = self._detectUsb
-					detectBluetooth = self._detectBluetooth
-					limitToDevices = self._limitToDevices
-				if detectUsb:
-					if self._stopEvent.isSet():
-						continue
-					for driver, match in getDriversForConnectedUsbDevices():
-						if self._stopEvent.isSet() or (self._limitToDevices and driver not in self._limitToDevices):
-							continue
-						if braille.handler.setDisplayByName(driver, detected=match):
-							return
-				if detectBluetooth:
-					if self._stopEvent.isSet():
-						continue
-					with self._btDevsLock:
-						if self._btDevs is None:
-							btDevs = list(getDriversForPossibleBluetoothDevices())
-							# Cache Bluetooth devices for next time.
-							btDevsCache = []
-						else:
-							btDevs = self._btDevs
-							btDevsCache = btDevs
-					for driver, match in btDevs:
-						if self._stopEvent.isSet() or (self._limitToDevices and driver not in self._limitToDevices):
-							continue
-						if btDevsCache is not btDevs:
-							btDevsCache.append((driver, match))
-						if braille.handler.setDisplayByName(driver, detected=match):
-							return
-					if self._stopEvent.isSet():
-						continue
-					if btDevsCache is not btDevs:
-						with self._btDevsLock:
-							self._btDevs = btDevsCache
+		if self._queuedFuture:
+			# This will cancel a queued scan (i.e. not the currently running scan, if any)
+			# If this future belongs to a scan that is currently running or finished, this does nothing.
+			self._queuedFuture.cancel()
+
+	# C901 '_bgScan' is too complex
+	# Note: when working on _bgScan, look for opportunities to simplify
+	# and move logic out into smaller helper functions.
+	def _bgScan(  # noqa: C901

```

Comment:
This comment is unresolved

Reply:
Ah, sorry, I missed this one.
I don't see how we could add unit tests here easily. The whole concept of auto detection of braille displays relies on hardware being connected and disconnected, initializing displays, etc. We could possibly write a fake display driver, but it would be pretty difficult to inject it into the detection logic that is hardware oriented. I feel that would be way beyond scope.

Reply:
@leonardder Although unit tests would be great, and certainly something work towards. The comment was referring to cyclomatic complexity failure (C901) from the lint step. See https://www.flake8rules.com/rules/C901.html. Essentially, the function has too many paths. This makes the purpose/outcomes of the function hard to understand, thus hard to maintain. Given you've had to change this function significantly, you must have taken the time to understand it. Now is a great time to leverage the understanding you have and re-write the function in smaller composable pieces that can be understood individually.

Reply:
Ah, great. Sorry I misunderstood, I'll have a look.